Gems/Atom/RHI/DX12/Code/Source/RHI/TileAllocator.cpp

```diff
@@ -0,0 +1,226 @@
+/*
+ * Copyright (c) Contributors to the Open 3D Engine Project.
+ * For complete copyright and license terms please see the LICENSE at the root of this distribution.
+ *
+ * SPDX-License-Identifier: Apache-2.0 OR MIT
+ *
+ */
+
+#include <RHI/Device.h>
+#include <RHI/TileAllocator.h>
+
+namespace AZ
+{
+    namespace DX12
+    {
+        namespace
+        {
+            const static bool TileAllocatorOutputDebugInfo = false;
+        }
+
+        void TileAllocator::Init(const Descriptor& descriptor, HeapAllocator& heapAllocator)
+        {
+            AZ_Assert(descriptor.m_getHeapMemoryUsageFunction, "You must supply a valid function for getting heap memory usage.");
+
+            m_descriptor = descriptor;
+            m_heapAllocator = &heapAllocator;
+
+            m_tileCountPerPage = m_heapAllocator->GetPageSize()/descriptor.m_tileSizeInBytes;
+        }
+
+        uint32_t TileAllocator::AllocateFromFreeList(uint32_t tileCount, AZStd::vector<HeapTiles>& output)
+        {
+            uint32_t allocatedTileCount = 0;
+
+            while ( !m_freeList.empty() && allocatedTileCount < tileCount)
+            {
+                uint32_t neededTileCount = tileCount - allocatedTileCount;
+                auto freeListItr = m_freeList.begin();
+                auto heap = *freeListItr;
+                auto itr = m_pageContexts.find(heap);
+
+                if (itr != m_pageContexts.end())
+                {
+                    uint32_t allocated = 0;
+                    auto tilesGroups = itr->second.TryAllocate(neededTileCount, allocated);
+
+                    if (itr->second.GetFreeTileCount() == 0)
+                    {
+                        m_freeList.erase(freeListItr);
+                    }
+
+                    if (allocated == 0)
+                    {
+                        AZ_Assert(false, "Implementation error: heap page in free list doesn't have tiles available");
```

Comment:
The AllocateFromFreeList function should always allocate successfully since it checked total free count before it's called. I am updating the function to return void. 

Any asserts in the function is to catch implementation issues.